### PR TITLE
Add chain break methods that use numpy arrays

### PR DESCRIPTION
**Pull request by arcondello**
Partially address #222 by providing an alternative path.

### CODE REVIEW

**Comment by arcondello** [comment]
Note: this is a backward compatibility breaking change
Closes #222 
Closes #177 
Closes #189

File: dimod/embedding/chain_breaks.py [hunk 1]
```diff
@@ -13,299 +13,368 @@
...
+        raise ValueError("expected samples to be a numpy 2D array")
 
-        if len(chain) == 1:
-            continue
+    num_samples, num_variables = samples.shape
+    num_chains = len(chains)
 
-        # there must be a better way
-        unbroken = np.array((samples_matrix[:, chain] == samples_matrix[:, v]).all(axis=1)).flatten()
+    broken = np.zeros((num_samples, num_chains), dtype=bool, order='F')
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/embedding/chain_breaks.py]: Slightly faster to overwrite the columns later on.

[the remark this answers is not in the document]
Comment (randomir) [reply dimod/embedding/chain_breaks.py]: Never mind, I see you're updating one column at a time. Makes sense for internal use. Btw, do you see users iterating over chains, or samples?

[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/embedding/chain_breaks.py]: Some of both depending on their use case.

File: dimod/embedding/chain_breaks.py [hunk 2]
```diff
@@ -13,299 +13,368 @@
...
-        >>> next(dimod.embedding.discard(samples, embedding), 'No sample')  # doctest: +SKIP
-        {'a': 1, 'b': 0}
-        >>> samples = {0: 1, 1: 0, 2: 0}
-        >>> next(dimod.embedding.discard(samples, embedding), 'No sample')
-        'No sample'
+        >>> import numpy as np
+        ...
+        >>> chains = [(0, 1), (2,)]
+        >>> samples = np.array([[1, 1, 0], [1, 0, 0]], dtype=np.int8)
+        >>> unembedded, idx = dimod.embedding.discard(samples, chains)
```
Comment (randomir) [inline dimod/embedding/chain_breaks.py]: Worth considering different/clearer namespace for chain resolution methods.

**Pull request merged**